File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long answer = 0;
        long start = 0;
        Arrays.sort(times);
        long end = times[times.length-1]*(long)n;
        
        while(start<=end){
            long mid = (start+end)/2;
            
            long count=0;
        
            for(int i=0;i<times.length;i++){
                count+=(mid/times[i]);
            }
            
            if(count<n){
              start = mid+1;  
            }
            else{
                end = mid-1;
                answer = mid;
            }
        }
        return answer;
    }
    
    public static int getCount(int[] times,long mid){
        int count=0;
        
        for(int i=0;i<times.length;i++){
            count+=(mid/times[i]);
        }
        
        return count;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long getCount(long long time,vector<int> times){
    
    long long count=0;

    for(int i=0;i<times.size();i++){
        
        count+=(time/(long long)times[i]);
    }
    
    return count;
}


long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    sort(times.begin(),times.end());
    
    long long start = 0;
    long long end = n * (long long)times.back();

    
    while(start<=end){
        long long mid = (start+end)/2;
        
        if(getCount(mid,times)>=n){
            end = mid-1;
            answer = mid;
        }
        else{
            start = mid+1;
        }
    }
    
    return answer;
}